Structured metadata support

RTLcoil opened the pull request.

RTLcoil commented:
@strausr The above PR is a work in progress.

Could you explain how to properly sign Upload API requests containing structured metadata? 

If you look at this [skipped test](https://github.com/cloudinary/cloudinary_npm/blob/f9c490e14d3660f4fa5659defdf07ec8d3d839b6/test/structured_metadata_spec.js#L150) and allow it to run you will get an `Invalid Signature` error. How do we properly sign the request?

strausr commented on lib/api.js:
```diff
@@ -498,3 +498,38 @@ exports.update_resources_access_mode_by_ids = function update_resources_access_m
   return updateResourcesAccessMode(access_mode, "public_ids[]", ids, callback, options);
 };
+
+exports.create_metadata_field = function create_metadata_field(metadata, callback, options = {}) {
```
Can you please change the callback to match v2 signature

RTLcoil replied:
@strausr I am not sure what you mean. Could you explain this please?

RTLcoil replied:
@strausr a possibly related question:
Should all the new functions be `1` or `2` in `lib/v2/api.js` and `lib/v2/uploader.js`? What does it mean?

strausr commented on lib/api.js:
```diff
@@ -498,3 +498,38 @@ exports.update_resources_access_mode_by_ids = function update_resources_access_m
+  return call_api("put", "metadata_fields/" + external_id, params, callback, options);
+};
+
+exports.update_metadata_field_datasource = function update_metadata_field_datasource(external_id, callback, options = {}) {
```
Metadata should be sent as payload and not taken from the options

strausr commented on lib/api.js:
```diff
@@ -498,3 +498,38 @@ exports.update_resources_access_mode_by_ids = function update_resources_access_m
+  return call_api("get", "metadata_fields/" + external_id, {}, callback, options);
+};
+
+exports.update_metadata_field = function update_metadata_field(external_id, callback, options = {}) {
```
Metadata should be sent as payload and not taken from the options

strausr commented on lib/api.js:
```diff
@@ -498,3 +498,38 @@ exports.update_resources_access_mode_by_ids = function update_resources_access_m
 ) {
   return updateResourcesAccessMode(access_mode, "public_ids[]", ids, callback, options);
 };
+
```
Can you please add DOCStrings to the methods?

strausr commented on lib/api.js:
```diff
@@ -498,3 +498,38 @@ exports.update_resources_access_mode_by_ids = function update_resources_access_m
+  return call_api("put", "metadata_fields/" + external_id + "/datasource", params, callback, options);
+};
+
+exports.delete_entries_field_datasource = function delete_entries_field_datasource(external_id, callback, options = {}) {
```
Method should be called delete_datasource_entries. Should pass in external_ids as payload rather than taken from options

strausr commented on test/structured_metadata_spec.js:
```diff
@@ -0,0 +1,160 @@
+const expect = require("expect.js");
```
Each test should be stand-alone and should not be based on previous tests

strausr commented on test/structured_metadata_spec.js:
```diff
@@ -0,0 +1,160 @@
+const expect = require("expect.js");
+const cloudinary = require("../cloudinary");
+const helper = require("./spechelper");
```
Need to finish the tests for upload, update, and explicit. This means adding functionality to the uploader. You can take a look at the implementation updateMetadata in Java

strausr commented on test/structured_metadata_spec.js:
```diff
@@ -0,0 +1,160 @@
+      expect(result).not.to.be.empty();
+    });
+  });
+  it.skip("should upload image with metadata option", function () {
```
Why is this test marked as skip? Also, the URL used for this test should not rely on a temporary cloud

RTLcoil replied:
This test failed and was thus marked `skip` because the metadata feature flag wasn't activated for our cloud.

It has since been activated and uploading works.
We also fixed what you asked for here.

RTLcoil commented:
Hey @strausr 

Thanks for the review. We implemented everything you suggested and then some.

There are still a few improvements I would like to get done on the tests, but the main code is ready and I wanted to get this to you ASAP as I know time is a factor.

BTW - I still have one open question to you. What did you mean in 
https://github.com/cloudinary/cloudinary_npm/pull/312#pullrequestreview-318349656 ?

RTLcoil commented:
Pull request updated.

* Updated private variable names and docstrings to match new reference implementation.
* Added missing function - `restore_metadata_field_datasource`
* Vastly improved tests.

strausr commented on test/spechelper.js:
```diff
@@ -11,7 +11,7 @@ const https = require('https');
 const cloudinary = require("../cloudinary");
```
Please change file name to spec_helper

strausr commented on test/structured_metadata_spec.js:
```diff
@@ -0,0 +1,356 @@
+          external_id: EXTERNAL_ID_CREATE,
+          label: LABEL_STRING_2,
+          type: "string",
+          default_value: "blue",
```
Should align test values to match PHP. Please remove default_value

strausr commented on test/structured_metadata_spec.js:
```diff
@@ -0,0 +1,356 @@
+const helper = require("./spechelper");
+
+const TEST_ID = helper.TEST_ID;
+const TEST_TAG = helper.TEST_TAG;
```
There are many tests missing or incomplete. For example, add test for Test creating an integer, enum, date, set metadata filed. Please align tests to match PHP implementation- cloudinary_php/pull/198
